File: go/bob/bob.go
/*
Package bob provides a Hey method to give a response from bob to a message.
*/
package bob

import (
	"regexp"
	"strings"
)

// Hey returns a response from bob.
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)
	if remark == "" {
		return "Fine. Be that way!"
	}
	if allCaps(remark) && question(remark) {
		return "Calm down, I know what I'm doing!"
	}
	if allCaps(remark) {
		return "Whoa, chill out!"
	}
	if question(remark) {
		return "Sure."
	}
	return "Whatever."
}

// allCaps determines if remark is a yelling.
func allCaps(remark string) bool {
	m, _ := regexp.Match("[a-z]", []byte(remark))
	M, _ := regexp.Match("[A-Z]", []byte(remark))
	return !m && M
}

// allCaps determines if remark is a question.
func question(remark string) bool {
	return remark[len(remark)-1] == '?'
}

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn remove-unword-chars [s]
  (apply str (filter (fn [c] (Character/isLetter c)) s)))

(defn shouting? [s]
  (let [clean-s (remove-unword-chars s)]
    (and (= clean-s (str/upper-case clean-s)) (> (count clean-s) 0))))

(defn questioning? [s]
  (str/ends-with? s "?"))

(defn nothing? [s]
  (str/blank? s))

(defn response-for [s]
    (cond
      (nothing? s) "Fine. Be that way!"
      (and (questioning? s) (shouting? s)) "Calm down, I know what I'm doing!"
      (questioning? s) "Sure."
      (shouting? s) "Whoa, chill out!"
      :else "Whatever.")
  )
